### DISCUSSION TITLE
Multiple init for separated modules

### DISCUSSION THREAD

**Discussion by adrien2p** [mikro-orm/mikro-orm#4552] [General]
**Describe the bug**
When using Mikro.init multiple times for separate module with each their entities, it seems that the metadata get lost in the process.

**Stack trace**
Entity metadata not found from the MetadataStorage

**To Reproduce**

The connection builder helper
```ts
export async function createConnection(
  database: ModuleServiceInitializeOptions["database"],
  entities: any[]
) {
  const schema = database.schema || "public"
  const orm = await MikroORM.init<PostgreSqlDriver>({
    discovery: { disableDynamicFileAccess: true },
    entities,
    debug: database.debug ?? process.env.NODE_ENV?.startsWith("dev") ?? false,
    baseDir: process.cwd(),
    clientUrl: database.clientUrl,
    schema,
    driverOptions: database.driverOptions ?? {
      connection: { ssl: true },
    },
    tsNode: process.env.APP_ENV === "development",
    type: "postgresql",
    migrations: {
      path: __dirname + "/../migrations",
    },
  })

  return orm
}
```

1. Init mikro ORM with entity A in module A
2. Init mikro ORM with entity B in module B
3. in module A try to use the repository for entity A
4. Receive the metadata not found error

**Expected behavior**
Shared global storage metadata

Also, I can see that the static property metadata of the MetadataStorage class contains all the entries, but not the instance property of the class.

**Versions**

| Dependency | Version |
| - | - |
| node | 18 |
| typescript | 4.9 |
| mikro-orm | 5.7.12 |
| your-driver | postgresql |

**Comment by B4nan** [comment]
You will need to provide complete repro for such things, there are tests using multiple ORM instances and things work just fine. [Here is one](https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/reusing-knex-client.test.ts) (and another one [here](https://github.com/mikro-orm/mikro-orm/blob/master/tests/issues/GH1226.test.ts)), tried to get a repository from both ORM instances and it works without any issues.

There is a global decorator metadata cache (as decorators are evaluated before the ORM init phase), but this cache is never cleared (you can only do that explicitly yourself).

**Reply by adrien2p** [reply]
thanks for the fast reply, we will check that on our side thanks for the reference to the tests. Will come back to you if we have further difficulties

**Reply by B4nan** [reply]
You could start with checking the actual metadata via `orm.getMetadata()` (at the place where it fails for you). Every ORM instance will have its own metadata copy, this is handled here:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/metadata/MetadataDiscovery.ts#L40

**Reply by adrien2p** [reply]
Yes I have already checked and when module A try to get the metadata it gets the ones from module A at the moment and therefore it fails but the discovery in itself works well. it fails from MetadataStore.get with validate true, at this moment the entity is not present as part of the metadata instance property. I will investigate some other places just to verify and check something else
<img width="1017" alt="Screenshot 2023-07-24 at 17 23 08" src="https://github.com/mikro-orm/mikro-orm/assets/25098370/eac9b9d6-26af-4edf-863c-693b7b91eb46">
<img width="1139" alt="Screenshot 2023-07-24 at 17 23 48" src="https://github.com/mikro-orm/mikro-orm/assets/25098370/1af3f806-a5db-43ce-9f95-c22f160ad78b">

**Reply by B4nan** [reply]
Also, how does the full stack trace look like? 

I can't imagine a case where you would get a missing metadata error for this, it sounds like you are using the wrong ORM instance, but you should see that from the `orm.getMetadata()` call.

**Reply by adrien2p** [reply]
After following your suggestion, I can see that both services from both modules are having the expected metadata. So it might comes from something else

<img width="977" alt="Screenshot 2023-07-24 at 17 28 33" src="https://github.com/mikro-orm/mikro-orm/assets/25098370/43b0553c-a706-416f-8175-e729ed06dd30">

and here the full stacktrace
```
MetadataError: Metadata for entity InventoryItem not found
    at Function.missingMetadata (/Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/node_modules/@mikro-orm/core/errors.js:178:16)
    at MetadataStorage.get (/Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/node_modules/@mikro-orm/core/metadata/MetadataStorage.js:54:42)
    at EntityFactory.create (/Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/node_modules/@mikro-orm/core/entity/EntityFactory.js:27:36)
    at SqlEntityManager.create (/Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/node_modules/@mikro-orm/core/EntityManager.js:863:41)
    at /Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/packages/inventory/src/repositories/inventory-item.ts:68:44
    at Array.map (<anonymous>)
    at InventoryItemRepository.create (/Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/packages/inventory/src/repositories/inventory-item.ts:67:24)
    at InventoryItemRepository.<anonymous> (/Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/packages/utils/src/modules-sdk/decorators/inject-transaction-manager.ts:26:37)
    at step (/Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/packages/utils/dist/modules-sdk/decorators/inject-transaction-manager.js:33:23)
    at Object.next (/Users/adriendeperetti/Documents/02-personnel/01-projects/medusa/packages/utils/dist/modules-sdk/decorators/inject-transaction-manager.js:14:53)
```

**Reply by adrien2p** [reply]
I suspect that is has to do something with one of our scenario of shared cradle for our DI. I ll come back to you to confirm that. Thank you very much for your time

**Reply by B4nan** [reply]
I think you have a context issue, and a wrong EM instance is being resolved. How do you set up the request context?

**Reply by adrien2p** [reply]
There is no request context here, we have multiple modules loaded as part of a core, each module can run independently with their own connection or with the same connection. The core receive the request and orchestrate the call to the different modules. The module from the entry point (public api) will use his own connection to get the manager and fork it. Each module can have their own orm and it is (will be) changeable. The problem I think comes from one of the loading part so i will figure this out but i think i might have found the issue, just need to confirm. I think that we can close this one for now 👍 thanks a lot for your guidance
